src/v/bytes/iobuf.h

```diff
@@ -389,14 +403,31 @@ iobuf::append(const uint8_t* src, size_t len) {
     }
     append(std::make_unique<fragment>(std::move(b)));
 }
+
 /// appends the contents of buffer; might pack values into existing space
-inline void iobuf::append(iobuf o) {
-    while (!o._frags.empty()) {
-        fragment* f = &o._frags.front();
-        o._frags.pop_front();
-        append(f->share());
-        details::dispose_io_fragment(f);
+inline void iobuf::append(iobuf&& o) {
+    for (auto& f : o._frags) {
+        auto fsize = f.size();
+
+        if (unlikely(fsize)) {
+            continue;
+        }
+
+        if (try_copy_append(f.get(), fsize)) {
+            continue;
+        }
+
+        if (available_bytes() > 0) {
+            if (_frags.back().is_empty()) {
+                pop_back();
+            } else {
+                _frags.back().trim();
+            }
+        }
+
+        append(std::move(f).untrimmed_release());
```

Comment:
What's the reason for this change to "untrimmed"?

Reply:
The reason for using `untrimmed_release` instead of just `release`? Or instead of `share`? The switch from`share` to some sort of release is so it doesn't promote a single-owner buffer to a reference counted multiple owner buffer. And I was avoiding `io_fragment::release` since it allocates/copies that data to a new buffer in some cases and therefore isn't a drop-in for `io_fragment::share` in some cases.

Reply:
Note that I've renamed `untrimmed_release` to `unoptimized_release`